scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java

```diff
@@ -652,12 +679,17 @@ public final void processMessage(Record record) throws HandshakeException {
 							if (epoch == 0) {
 								handshakeMessages.add(handshakeMessage);
 							}
-							doProcessMessage(handshakeMessage);
+							recursionProtection.lock();
```

Comment:
With the async handling it gets required to process the buffered handshake messages also from `processPskSecretResult`, if that is called from  `DTLSConnector.processPskSecretResult`. If it is called from handshaker's `doProcessMessage` then this will fail with `expectMessage`, because the `statesIndex` is incremented on returning `doProcessMessage`. So a recursive call must be returned in order to process the buffer handshake messages with the normal processing.
I also considered to split `processPskSecretResult` so that only the external call calls `processNextMessages`. If that would be violated accidentally, that will cause a hard to find rare error. So I decided to mitigate the cause.

Reply:
(keep in mind that maybe I don't get the issue)

Why not extracting the removing the processNextMessages() from `processPskSecretResult` and call it only from `DTLSConnector.processPskSecretResult` ?

Reply:
Uups this is pretty much what you propose here : 
> I also considered to split processPskSecretResult so that only the external call calls processNextMessages

Reply:
I think this would be clearer.

> If that would be violated accidentally, that will cause a hard to find rare error. So I decided to mitigate the cause.

I'm not sure to get this point.

Reply:
Yes, I already tried that.
I can adapt the code accodingly, but I would keep the recursion protection. It 's really hard, it costs me more than two days of debugging ...

Reply:
Ok could you add a comment to say this should never happen and replace the trace by a warn ?

Reply:
and explain that the lock is only for debug purpose (and maybe temporary)

Reply:
In the current implementation, it happens normally. 
With adapting, I will change the logging.

Reply:
Yep I got it :-)

Reply:
I pushed the adapted implementation.